File: Top Interview Questions/322. Coin Change.java
// bottom-up
class Solution {
    public int coinChange(int[] coins, int amount) {
        int max = Integer.MAX_VALUE - amount;
        
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, max);
        dp[0] = 0;
        //System.out.println(dp[amount]);
        
        for (int i = 1; i <= amount; i++) {
            for (int j = 0; j < coins.length; j++) {
                if (coins[j] <= i) dp[i] = Math.min(dp[i], dp[i - coins[j]] + 1);
            }
        }
        return dp[amount] < max ? dp[amount] : -1;
    }
}

// top-down: important dp example
class Solution {
    public int coinChange(int[] coins, int amount) {
        if (amount <= 0) return 0;
        return coinChange(coins, amount, new int[amount]);
    }
    
    public int coinChange(int[] coins, int amount, int[] counter) {
        if (amount < 0) return -1;
        else if (amount == 0) return 0;
        if (counter[amount - 1] != 0) return counter[amount - 1];
        
        int min = Integer.MAX_VALUE;
        
        for (int coin : coins) {
            int temp = coinChange(coins, amount - coin, counter);
            if (temp >= 0 & temp < min) min = temp;
        }
        counter[amount - 1] = (min < Integer.MAX_VALUE ? min + 1 : -1);
        return counter[amount - 1];
    }
}

File: Related Topics/Dynamic Programming/322. Coin Change.py
# 322. Coin Change
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [amount + 1 for _ in range(amount + 1)]
        
        # base case
        dp[0] = 0
        
        # status
        for i in range(amount + 1):
            # choices
            for coin in coins:
                if i - coin < 0:
                    continue
                dp[i] = min(dp[i], dp[i - coin] + 1)
                
        return dp[amount] if dp[amount] != amount + 1 else -1